presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java

```diff
@@ -337,6 +337,19 @@ public void testFloats()
         assertFilterProject("reals[1] > 0.01", "count(*)");
     }
 
+    @Test
+    public void testInComplexTypes()
```

Comment:
Please add this test case as well to make sure that correct answer (`NULL`) is returned.

```
        StringBuilder query = new StringBuilder("select ROW(null_value) IN ( ");
        for (int i = 0; i < 32; i++) {
            query.append(String.format("ROW(%s), ", i));
        }
        query.append("ROW(32)) ");
        query.append("FROM (values(null)) as t (null_value)");
        assertQuery(query.toString(), "SELECT NULL");
```

The `ROW(NULL) IN (ROW(0), ...) ` creates a collision required to get the equals path executed.

Also I would suggest moving this test to `AbstractTestQueries`, as there's nothing Hive specific about it.

Also please add a comment describing what kind of black magic are we doing here.